# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [1]:
import pandas as pd

# Create the series
temperatures : pd.core.series.Series = pd.Series([72, 75, 68, 71, 73], index=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"])

# Print the series and its data type
print(temperatures)
print("\n")
print(temperatures.dtype)

Monday       72
Tuesday      75
Wednesday    68
Thursday     71
Friday       73
dtype: int64


int64


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [2]:
products_as_dict : dict = {"product" : ["Widget", "Gadget", "Doohickey"], "price" : [19.99, 24.99, 12.49], "quantity" : [100, 75, 150]}

products : pd.core.frame.DataFrame = pd.DataFrame(products_as_dict)

print(products)
print(f"\nThe DataFrame products has a size of {products.size}")

     product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150

The DataFrame products has a size of 9


#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [3]:
# Print the price as a Series
price_as_series : pd.core.series.Series = products["price"]
print(price_as_series)

# Print the product and quantity as a DataFrame
prod_quant_as_df : pd.core.frame.DataFrame = products[["product", "quantity"]]
print("\n")
print(prod_quant_as_df)

0    19.99
1    24.99
2    12.49
Name: price, dtype: float64


     product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

##### Your Answers

### Student Response:

1. When a DataFrame is created from a dictionary, the keys become column headers, and the values become data points.
2. `df\['price'\]` returns a Series because the index is one diminsional, while `df\[\['price'\]\]` returns a DataFrame because the index is two diminsional. When using double square brackets to index, you are asking pandas to return the dictionary with only those columns, while single square brackets asks to return a series of a given column.
3. Using a custom index, as opposed to numerical indices, is more useful that the default index in instances where you may wish to plot data against this custom index, or an instance where the string values are unique (e.g. a users email)

### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [4]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('./data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)

# Load the raw data
df_raw : pd.core.frame.DataFrame = pd.read_csv(csv_path)

# Explore the data
print(f"df_raw has shape {df_raw.shape}")

print("\n")

#Print each column name, and its data type
print("Columns and their types:")
column_dtype_pairs : str = ("\n").join([(f"{col} : {df_raw[col].dtype}") for col in df_raw.columns])
print(column_dtype_pairs)

# See the head of the DataFrame
print("\n")
print(df_raw.head())

print(f"\n{df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

df_raw has shape (100000, 13)


Columns and their types:
ride_id : object
rideable_type : object
started_at : object
ended_at : object
start_station_name : object
start_station_id : object
end_station_name : object
end_station_id : object
start_lat : float64
start_lng : float64
end_lat : float64
end_lng : float64
member_casual : object


            ride_id  rideable_type               started_at  \
0  67BB74BD7667BAB7  electric_bike  2024-09-30 23:12:01.622   
1  5AF1AC3BA86ED58C  electric_bike  2024-09-30 23:19:25.409   
2  7961DD2FC1280CDC   classic_bike  2024-09-30 23:32:24.672   
3  2E16892DEEF4CC19   classic_bike  2024-09-30 23:42:11.207   
4  AAF0220F819BEE01  electric_bike  2024-09-30 23:49:25.380   

                  ended_at         start_station_name start_station_id  \
0  2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave           bdd4c3   
1  2024-10-01 00:42:09.933                        NaN              NaN   
2  2024-10-01 00:23:18.647     St. Clair St & Erie St      

#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [5]:
# Reload the data
rides : pd.core.frame.DataFrame = pd.read_csv(csv_path, dtype=
                                              {"ride_id" : str,
                                              "rideable_type" : "category",
                                              "start_station_name" : "category",
                                              "start_station_id" : "category",
                                              "end_station_name" : "category",
                                              "end_station_id" : "category",
                                              "start_lat" : float,
                                              "start_long" : float,
                                              "end_lat" : float,
                                              "end_long" : float,
                                              "member_casual" : "category"}, 
                                              parse_dates= ["started_at", "ended_at"])

print(rides.info())

print("\nCategories Memory Usage:")
print(f"{rides.memory_usage(deep=True).sum() / 1024**2:.2f} MB")
print("\nRaw Memory Usage:")
print(f"{df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")
print("The categories instance is much smaller")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  object        
 1   rideable_type       100000 non-null  category      
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   category      
 5   start_station_id    89623 non-null   category      
 6   end_station_name    89485 non-null   category      
 7   end_station_id      89485 non-null   category      
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  category      
dtypes: category(6), datetime64[ns]

#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [6]:
# Determine the range of the starting dates
start_date_min = rides["started_at"].min()
start_date_max = rides["started_at"].max()
print(f"The starting dates range from {start_date_min} to {start_date_max}\n")

# Count the number of missing values in each column, then convert this Series into a DataFrame
missing_val_counts : pd.core.series.Series = rides.isna().sum()
missing_val_counts_df : pd.core.frame.DataFrame = missing_val_counts.to_frame(name='count')

# Add a column for percent missing
missing_val_counts_df["percentage"] = [((missing_val_counts_df.loc[idx, "count"]/len(rides[idx]))*100) for idx in missing_val_counts_df.index]

# Print the resulting DataFrame
print(missing_val_counts_df)

The starting dates range from 2024-09-30 23:12:01.622000 to 2024-10-31 23:54:02.851000

                    count  percentage
ride_id                 0       0.000
rideable_type           0       0.000
started_at              0       0.000
ended_at                0       0.000
start_station_name  10377      10.377
start_station_id    10377      10.377
end_station_name    10515      10.515
end_station_id      10515      10.515
start_lat               0       0.000
start_lng               0       0.000
end_lat                87       0.087
end_lng                87       0.087
member_casual           0       0.000


#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [7]:
# Get the trip duration in minutes
rides["trip_duration_min"] = ((rides["ended_at"] - rides["started_at"]).dt.total_seconds()/60)

# Use .describe() to display basic statistics of this Series
print(rides["trip_duration_min"].describe())

# Show the first few rows with the specified columns
print("\n")
print(rides[["started_at", "ended_at", "trip_duration_min"]].head(3))

# Set the index as "started_at" and print the first few rows
rides = rides.set_index("started_at")
print(f"\n\n\nThe index first is: {rides.index[0]}\n\n\n")
print(rides.head(3))

count    100000.000000
mean         16.144576
std          52.922539
min           0.006533
25%           5.489271
50%           9.423592
75%          16.407171
max        1499.949717
Name: trip_duration_min, dtype: float64


               started_at                ended_at  trip_duration_min
0 2024-09-30 23:12:01.622 2024-10-01 00:20:00.674          67.984200
1 2024-09-30 23:19:25.409 2024-10-01 00:42:09.933          82.742067
2 2024-09-30 23:32:24.672 2024-10-01 00:23:18.647          50.899583



The index first is: 2024-09-30 23:12:01.622000



                                  ride_id  rideable_type  \
started_at                                                 
2024-09-30 23:12:01.622  67BB74BD7667BAB7  electric_bike   
2024-09-30 23:19:25.409  5AF1AC3BA86ED58C  electric_bike   
2024-09-30 23:32:24.672  7961DD2FC1280CDC   classic_bike   

                                       ended_at      start_station_name  \
started_at                                                           

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

### Student Response:
1. Pandas assigned `started_at` and `member_casual` as `object` data types. This can be problomatic because it can be difficult to do operations on these ambigous data types. The datetime type and category type allowed easier operations, and memory efficiencies, respectively.
2. Station IDs were possibly obtained by computing the cryptographic hash of the station name, to guarantee a uniqe station ID per station.
3. Method chaining is the act of calling methods on objects returned by other methods. `df.isna().sum()` first computes a boolean mask of `df`, and then takes the sum of the resulting columns in the DataFrame, collapsing the DataFrame into a Series. This chain is a way to compute the number of missing data values per column.
4. This situation could be explained by riders leaving their rideables _near_ but not within the geofence boundary of the station. To handle affected rows, you could assign stations to rides if the coordinates are within some tolerance distance of the station coordinates.

#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

### Student Response
2a.3 used 64MB of memory, while 2b.3 used approximately 13MB of memory. The change is driven by the use of categories. Categories store each value only once and reuse duplicates, saving memory for repeat values. The value reported from the use of `df.info()` only includes the memory impact of pointers to values, and not the values themselves. If data size is a concern, use the value that includes the values themselves, and not just the pointers. 

Assuming 30% overhead for various programs and OS PIDs, there are 11.2 GB of memory remaining for use.

**For the optimistic case:**
We can solve for maxmium data size by solving 5x = 11.2, for x. Here, x = 2.24 GB

**For the pessimistic case:**
We can solve for maxmium data size by solving 10x = 11.2, for x. Here, x = 1.12 GB

I would reccomend working with datasets no larger than 1.12 GB on a machine with 16 GB of RAM. This allows plenty of overhead in the event that another OS process or program starts using more RAM.

### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [8]:
# Compute the number of trips taken by members usiung electric bikes
trips_with_members : pd.core.frame.DataFrame = rides["member_casual"] == "member"
trips_with_ebikes : pd.core.frame.DataFrame = rides["rideable_type"] == "electric_bike"

count_rides_members_ebikes : int = len(rides[(trips_with_ebikes) & (trips_with_members)])
print(f"There were {count_rides_members_ebikes} trips taken by members on electric bikes.")
print(f"This represents {count_rides_members_ebikes/len(rides):.2%} of all trips.")

# Compute the number of trips that started in one of the specified stations
tgt_stations : list = ["Streeter Dr & Grand Ave" ,"DuSable Lake Shore Dr & Monroe St", "Kingsbury St & Kinzie St"]
count_trip_started_in_specified : int = len(rides[rides["start_station_name"].isin(tgt_stations)])
print(f"{count_trip_started_in_specified} trips began in one of the specified stations.")

There were 33121 trips taken by members on electric bikes.
This represents 33.12% of all trips.
2702 trips began in one of the specified stations.


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [16]:
# Create a column called is_weekend to have a boolean value of true if the start time of the trip is on a weekend
weekend_bool_mask : pd.core.frame.DataFrame = rides.index.dayofweek.isin([5,6])
rides["is_weekend"] = weekend_bool_mask
rides["is_weekend"] = rides["is_weekend"].astype("category")

# Create a column called time_of_day that categorizes each ride based on the time of day

# Initialize on Midday
rides["time_of_day"] = "Midday"

# Mask the Morning Rush case
rides.loc[rides.index.hour.isin([7, 8, 9]), "time_of_day"] = "Morning Rush"

# Mask the Evening Rush case
rides.loc[rides.index.hour.isin([16, 17, 18]), "time_of_day"] = "Evening Rush"

# Convert the column to a category column
rides["time_of_day"] = rides["time_of_day"].astype("category")

print("Distribution of time_of_day")
print(rides["time_of_day"].value_counts())

Distribution of time_of_day
time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64


#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [23]:
# Use the query() method to find the trips that meet all of the provided criteria
rides_meeting_criteria : pd.core.frame.DataFrame = rides.query("member_casual == 'casual' and is_weekend == True and trip_duration_min > 20 and rideable_type == 'electric_bike'", engine='python')

# Of the resulting trips, find the average of the duration
avg_duration_meet_criteria : float = rides_meeting_criteria["trip_duration_min"].mean()

print("REPORT:")
print(f"Trips matching criteria: {len(rides_meeting_criteria)}")
print(f"Percent of total trips matching criteria: {len(rides_meeting_criteria)/len(rides):.2%}")
print(f"Average duration of trips meeting criteria: {avg_duration_meet_criteria:.2f} minutes")

REPORT:
Trips matching criteria: 1501
Percent of total trips matching criteria: 1.50%
Average duration of trips meeting criteria: 40.37 minutes


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [33]:
# Select first 10 trips with member_casual, rideable_type, and trip_duration_min columns only
first_10_3cols : pd.core.frame.DataFrame = rides[["member_casual", "rideable_type", "trip_duration_min"]].iloc[:10]

# Select trips from Oct. 15-17, with the same three cols
october_trips : pd.core.frame.DataFrame = rides.loc['2024-10-15':'2024-10-17', ["member_casual", "rideable_type", "trip_duration_min"]]

# Report the number of trips that occured in this date range
print(f"{len(october_trips)} trips occurred from Oct. 15 to Oct. 17th.")

7235 trips occurred from Oct. 15 to Oct. 17th.


#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

### Student Response
1. Using `isin()` is not only more readable, but allows the user to build a dynamic list of values to test against (using list comprehension on some values with perhaps even a set of logical conditions. This method also allows the user to write code that is not contained in a single, very long line.
2. If we did not set a default, or initialize all values as Midday, we would have to have a very long list of elements to test each values memebership against. There also is not a clean mechanic for "else" logic in boolean assignement.
3. The `query()` method is better for several logical tests in conjunction. However, boolean indexing might be preferable, despite being more verbose, in instances where compound conditions, or more complex compound conditions are tested for (e.g. (A and B) or C if C in someList). Boolean indexing might also be prefered when variables represent the same name as column names, or a dynamic condition is needed.


#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

### Student Response

Bracket notation, while technically viable, is best reserved for column selection because it is more readable. Being consistent with use of bracket notation produces more readable code. For example, `df[0:5]` appears to be an index label slice, however, if the index was a date, as in this assignment, the fact that this expression conducts positon indexing produces ambigous results. Using `loc[]` for label based indexing solves this problem by explicitly conveying to the reader that label based indexing is being used. Similarly `iloc[]` usage conveys explicitly that position based indexing is being used. The differentiator is that by sticking to the prescribed convention above, intent is always clear and consistent.

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [67]:
# Get a Series of the value_counts() and find the max
index_hours_counts : pd.core.frame.DataFrame = (rides.index.hour).value_counts()
print(f"The peak hour is hour {index_hours_counts.idxmax()} with {index_hours_counts.max()} rides.")

# Extract the day name from the index and find the busiest day
index_days_counts : pd.core.frame.DataFrame = rides.index.day_name().value_counts()
print(f"The peak day is day {index_days_counts.idxmax()} with {index_days_counts.max()} rides.")

# Sort the index days counts by day name
day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
print("\nSorted Results:")
print(index_days_counts.sort_index(key=lambda day : pd.Categorical(day, categories=day_order)))

The peak hour is hour 17 with 10574 rides.
The peak day is day Wednesday with 16513 rides.

Sorted Results:
started_at
Monday       11531
Tuesday      14970
Wednesday    16513
Thursday     16080
Friday       13691
Saturday     14427
Sunday       12788
Name: count, dtype: int64


#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [85]:
# Calculate average trip duration by rider type. Find the group that takes the longer trips on average.
rider_type_avg_duration : pd.core.series.Series = rides.groupby('member_casual', observed=False)['trip_duration_min'].mean()
print(f"The {rider_type_avg_duration.idxmax()} rider takes the longest trips on average with an average trip duration of {rider_type_avg_duration.max():.2f} minutes.")

# Calculate the average trip duration by bike type.
bike_type_avg_duration : pd.core.series.Series = rides.groupby('rideable_type', observed=False)['trip_duration_min'].mean()
print(f"The {bike_type_avg_duration.idxmax()} rideable takes the longest trips on average with an average trip duration of {bike_type_avg_duration.max():.2f} minutes.")

# Count the number of trips by rider type - groupby().size() method
trips_by_rider_type_method1 : pd.core.series.Series = rides.groupby('member_casual', observed=False)['trip_duration_min'].size()
print("\nTrips by rider type using the .groupby().size() method:")
print(trips_by_rider_type_method1)

# Count the number of trips by rider type - value_counts() method
trips_by_rider_type_method2 : pd.core.series.Series = rides['member_casual'].value_counts()
print("\nTrips by rider type using the value_counts() method:")
print(trips_by_rider_type_method2)

# These methods produce the same result

The casual rider takes the longest trips on average with an average trip duration of 23.98 minutes.
The classic_bike rideable takes the longest trips on average with an average trip duration of 20.34 minutes.

Trips by rider type using the .groupby().size() method:
member_casual
casual    34686
member    65314
Name: trip_duration_min, dtype: int64

Trips by rider type using the value_counts() method:
member_casual
member    65314
casual    34686
Name: count, dtype: int64
34686


#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [95]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Filter for trips taken during the weekend and the rideable is the electric bike
weekend_ebike_trips : pd.core.frame.DataFrame = rides.query("is_weekend == True and rideable_type == 'electric_bike'", engine="python")

# Update the DataFrame to only include the first 1000 results
weekend_ebike_trips_sample : pd.core.frame.DataFrame = weekend_ebike_trips.iloc[:1000]

# Reset the index to include the datetime index back into a column
weekend_ebike_trips_sample.reset_index(inplace=True)

# Export specified columns to csv
weekend_ebike_trips_sample.to_csv(output_file, index=False, columns=["started_at", "ended_at", "member_casual", "trip_duration_min", "time_of_day"])

# Report the number of weekend ebike trips
print(f"Before filtering, there were {len(weekend_ebike_trips)} weekend ebike trips.")

Before filtering, there were 13026 weekend ebike trips.


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

### Student Response
1. `groupby()` is a Pandas operation that applies the "split-apply-combine" operation to a DataFrame, or Series object. First, Pandas splits data by values in the specified column. Then, a function is applied to each independant group (sum(), size(), mean(), etc.). Finally, the function results for each group are combined into a single data structure (DataFrame or Series).
2. The `value_counts()` method is more suited for determining the frequency of many unique values in a single column, while the `groupby().size()` method is more well suited for determining the frequency of multiple columns' various combinations.
3. Had we not reset the index before exporting, with index=False, the started_at column would have been lost, and not reported in the csv export.

#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [96]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

# Filter for interesting pattern: Long trips (>15 min) during rush hours 
# from lakefront stations, sorted by duration
lakefront_rush = (rides
    .loc[(rides.index.hour.isin([7, 8, 9, 16, 17, 18]))]
    .loc[(rides['start_station_name'].str.contains('Lake Shore|Lakefront', 
                                                    case=False, 
                                                    na=False))]
    .loc[rides['trip_duration_min'] > 15]
    .sort_values('trip_duration_min', ascending=False)
    .head(1000)
    .reset_index()
    [['started_at', 'ended_at', 'start_station_name', 'end_station_name',
      'member_casual', 'rideable_type', 'trip_duration_min']]
)

print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare file sizes
csv_size = os.path.getsize(csv_file) / 1024  # Convert to KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Compare load times
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Check data type preservation
# Note: CSV load without parse_dates loses datetime types
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)

print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)

Found 310 long rush-hour trips from lakefront stations

CSV file size: 40.57 KB
Pickle file size: 55.16 KB
Size difference: 14.59 KB

Load time comparison:
CSV:
986 μs ± 1.52 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Pickle:
410 μs ± 1.51 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Data types from CSV (without parse_dates):
started_at             object
ended_at               object
start_station_name     object
end_station_name       object
member_casual          object
rideable_type          object
trip_duration_min     float64
dtype: object

Data types from Pickle:
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name          category
end_station_name            category
member_casual               category
rideable_type               category
trip_duration_min            float64
dtype: object


After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


### Student Response
1. Method chaining makles code more readable because the reader can examine the overall option in a step-by-step manner. Furthermore, this makes debugging easier by being able to look at each method operation individually. This relates to seeing changes in git diff in the sense that git diff will show the change to each line (each method call) as opposed to changes in a single line only (a single .method1().method2() call. Heavy chaining can cause issues if a more central method call has an error, it can be difficult to detect, due to output transformation though other method calls.
2. Data types in the CSV are not maintained from the DataFrame, while the pickle retains the data type from the DataFrame. The category and datetime data types are the types that are retained by the pickle. This is significant for subsequent analysis because the datatime retention cuts down on duplicate processing to cast those values back into datetime vales. The retention of the category type is critical for subsequent analsysis because it reduces memory use by reusing reoccuring values.
3. The pickle file is slightly larger than the csv file, but loads in under half of the time. Pickle is best used when speed matters, when memory usage matters, or subsequent analysis with special types will occur. The CSV file is the better choice when a human readable file is required, or you need to share the file with someone who may use a different language to perform subsequent analysis.

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

### Student Response
1. The biggest conceptual shift when moving from NumPy arrays to Pandas DataFrames is indexing. While NumPy offers a plethora of methods to index data, Pandas indexing relies more on column headers, and potentially complex index values to index the data structures. The most difficult Pandas concept is probably datetime operations. I struggled with identifying when to use the dt class, and when not to. Pandas documentation helps, and using an LLM to help decipher when to use which classmethod to call on which objects allowed me to work through this issue.
2. Working with real CSV data differs from working with NumPy arrays because the CSV data is heterogenous, which means a decent amount of setup is required to ensure data types are all being loaded in the right way, and the most efficient way (e.g. using category data types). Pandas is well-suited for data analysis when comapred to NumpPy since NumPy arrays are homogenous. Data is rarely homogenous across a data set, so there are limitations to what can be done with pure NumPy.
3. I think the `.query()` method is most useful to my own data work. I often analyze data at work where 5+ conditions must be tested for, each one affecting the rest. Using this query method will allow me to conduct these checks in a single line. I feel that my Pandas familiarity is about a 7/10. I believe that this will increase with the completion of the class project. Furthermore, when time allows, I plan to examine the Pandas class code for DataFrame and Series to understand which methods can be performed on which data types, and how they are to be used for each type.
4. I spent about 5 cumulative hours on this assignment. Sometimes, I worked on this assignment late at night after work, so I think I could have done it in about 3 hours if I worked on it during the day, for what it is worth. The most helpful parts of these assignments to me are the coding portions. I would like to see these assignments have more analysis and fewer reflection questions. I find the content derived from the reflection questions to be quite helpful, though. Perhaps some of the simpler conclusions from these questions can be conveyed in the problem statments (i.e. Use this method, now use this method, notice that method 1 is more readable/more concise/etc. than method 2) similar to what was done in the first homework.